arborist/engine.go

```diff
@@ -0,0 +1,459 @@
+type AuthEngine struct {
+	// The base role of the tree. The root role is kept empty aside from its
+	// subroles, which form the roots of trees in the forest.
+	root_role *Role
+
+	// Keep track of the role names used. This way, the engine searches for
+	// roles by name in constant time, and can also check in constant time that
+	// new roles have unique names. Make sure to add new roles to the map.
+	roles map[string]*Role
+
+	// Keep track of the resources used, by ID.
+	resources map[string]*Resource
+
+	// Keep track of existing permissions by ID.
+	permissions map[string]*Permission
+
+	// Keep track of services by ID.
+	services map[string]*Service
+}
+
+// NewAuthEngine creates a new engine with a blank role tree (containing just
+// the root role).
+func NewAuthEngine() (*AuthEngine, error) {
+	root_role, err := NewRole("root")
+	if err != nil {
+		return nil, err
+	}
+
+	roles := make(map[string]*Role, 1)
+	roles["root"] = root_role
+
+	resources := make(map[string]*Resource)
+
+	permissions := make(map[string]*Permission)
+
+	services := make(map[string]*Service)
+
+	engine := &AuthEngine{
+		root_role:   root_role,
+		roles:       roles,
+		resources:   resources,
+		permissions: permissions,
+		services:    services,
+	}
+
+	return engine, nil
+}
+
+// ListRoleNames returns just the names of all the roles stored in the engine.
+func (engine *AuthEngine) ListRoleNames() []string {
+	var result []string = make([]string, 0)
+	for role_name := range engine.roles {
+		result = append(result, role_name)
+	}
+	return result
+}
+
+// ListServiceNames returns just the names of the services that have been
+// stored.
+func (engine *AuthEngine) ListServiceNames() []string {
+	var result []string = make([]string, 0)
+	for service_name := range engine.services {
+		result = append(result, service_name)
+	}
+	return result
+}
+
+// ListResourceNames returns a slice of all the names of the resources stored in
+// the engine.
+func (engine *AuthEngine) ListResourceNames() []string {
+	var result []string
+	for resource_name := range engine.resources {
+		result = append(result, resource_name)
+	}
+	return result
+}
+
+// Return a list of references to all the resources created in this engine.
+func (engine *AuthEngine) allResources() []*Resource {
+	var result []*Resource
+	for _, resource := range engine.resources {
+		result = append(result, resource)
+	}
+	return result
+}
+
+// Look up a particular resource in a particular service.
+func (engine *AuthEngine) findResource(resourceID string) *Resource {
+	return engine.resources[resourceID]
+}
+
+// Return a list of references to ALL the roles in the engine (basically
+// flattening the tree).
+func (engine *AuthEngine) allRoles() []*Role {
+	return engine.root_role.allSubroles()
+}
+
+// findRole looks up the first role in the tree satisfying the predicate
+// function.
+func (engine *AuthEngine) findRole(predicate func(Role) bool) (*Role, error) {
```

Comment:
unused function?

Reply:
I have high expectations that this will be used if for instance the API supports filtering for roles by tags or type

Reply:
(for posterity: probably filtering via query string args)